File: java/DisjointSet.java
/*
  Disjoint Set Union (Union-Find) in Java
  ----------------------------------------
  Features:
  - Path Compression (optimizes find operation)
  - Union by Rank (balances tree height)
  
  Time Complexity:
  - find(): O(α(N)) ~ constant
  - union(): O(α(N)) ~ constant
*/

class DisjointSet {
    int[] parent;
    int[] rank;

    // Constructor: initialize DSU for 'n' elements
    DisjointSet(int n) {
        parent = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            rank[i] = 0;
        }
    }

    // Find with path compression
    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]); // compress path
        }
        return parent[x];
    }

    // Union by rank
    void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX == rootY) return; // already in same set

        if (rank[rootX] < rank[rootY]) {
            parent[rootX] = rootY;
        } else if (rank[rootX] > rank[rootY]) {
            parent[rootY] = rootX;
        } else {
            parent[rootY] = rootX;
            rank[rootX]++;
        }
    }

    // Example usage
    public static void main(String[] args) {
        DisjointSet dsu = new DisjointSet(5);

        dsu.union(0, 2);
        dsu.union(4, 2);
        dsu.union(3, 1);

        // Should print "Connected"
        if (dsu.find(4) == dsu.find(0)) {
            System.out.println("Connected");
        } else {
            System.out.println("Not Connected");
        }

        // Should print "Not Connected"
        if (dsu.find(1) == dsu.find(0)) {
            System.out.println("Connected");
        } else {
            System.out.println("Not Connected");
        }
    }
}

File: python/disjoint_set.py
class DSU:
    def __init__(self, n):
        # Initially, each node is its own parent
        self.parent = [i for i in range(n)]
        self.rank = [0] * n  # Rank for union by rank

    def find(self, x):
        if self.parent[x] != x:
            # Path compression
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        xroot = self.find(x)
        yroot = self.find(y)

        if xroot == yroot:
            return  # Already in the same set

        # Union by rank
        if self.rank[xroot] < self.rank[yroot]:
            self.parent[xroot] = yroot
        elif self.rank[xroot] > self.rank[yroot]:
            self.parent[yroot] = xroot
        else:
            self.parent[yroot] = xroot
            self.rank[xroot] += 1

# Example usage
if __name__ == "__main__":
    n = 5
    dsu = DSU(n)

    # Merge sets
    dsu.union(0, 1)
    dsu.union(1, 2)
    dsu.union(3, 4)

    # Find representatives
    for i in range(n):
        print(f"Parent of {i}: {dsu.find(i)}")
